### PR TITLE
Allow expressions for extension_sql and extension_sql_file macros

### PR DESCRIPTION
**Pull request by JamesGuthrie**
This builds on the open PR: https://github.com/zombodb/pgx/pull/303, allowing for arbitrary expressions in the `extension_sql_file` macro.

### CODE REVIEW

**Comment by JamesGuthrie** [comment]
I'd like to note that one of the main caveats with the change to `extension_sql_file` is that it requires setting the `name` attribute. This stems from the fact that we can't evaluate the expression to determine the name.

It's also possibly not intuitive that it only works for a certain subset of expressions and macros, namely those which are fully resolved at compile time.

File: pgx-examples/custom_sql/src/lib.rs [hunk 1]
```diff
@@ -67,12 +67,14 @@ extension_sql!(
 
 // `extension_sql_file` does the same as `extension_sql` but automatically sets the `name` to the
 // filename (not the full path).
-extension_sql_file!("../sql/single.sql", requires = ["single_raw"]);
+extension_sql_file!("../sql/single.sql", name = "single", requires = ["single_raw"]);
```
Comment (Hoverbear) [inline pgx-examples/custom_sql/src/lib.rs:70]: I see you added a test for `extension_sql` for this feature, could make one of the examples above do it too?

**Comment by JamesGuthrie** [comment]
> I'm wondering if there is a way we can preserve the auto-naming if the user passes a `syn::LitStr` instead of an `syn::Expr`...

I was wondering the same. I'm not familiar enough with proc macros or syn to know if that is possible.

File: pgx-examples/custom_sql/src/lib.rs (hunk 1 above)
Comment (JamesGuthrie) [reply pgx-examples/custom_sql/src/lib.rs:70]: Ya, I will add that. Interestingly I was unable to get `extension_sql_file` to work from the tests in `pgx-tests`. I couldn't figure out why though.

Comment (Hoverbear) [reply pgx-examples/custom_sql/src/lib.rs:70]: Hmm you're right, if I add

```rust

extension_sql_file!(
    concat!("../../../pgx-examples/custom_sql/sql/single.sql"),
    name = "example_file",
);
```

To the test file it breaks on 

```
[2022-01-27 10:55:38.814 PST] [18676] [61f2eaaa.48f4]: LOG:  statement: ROLLBACK

relation "extension_sql_dynamic" does not exist
Postgres location: parse_relation.c:1177
Rust location: <unknown>

', pgx-tests/src/framework.rs:120:21
```

Comment (Hoverbear) [reply pgx-examples/custom_sql/src/lib.rs:70]: If I add that, actually, all the `extension_sql` blocks disappear too??? So odd.

Comment (JamesGuthrie) [reply pgx-examples/custom_sql/src/lib.rs:67]: done

**Comment by Hoverbear** [comment]
This PR is causing some strangeness on `pgx-tests`'s sql generation.

I attempted to add the following to `pgx-tests/src/tests/extension_sql_tests.rs`:

```rust
extension_sql_file!(
    "../../sql/example.sql",
    name = "example_file",
);
```

Where `../../sql/example.sql` contained `CREATE TABLE example();`.

Once I did this all `extension_sql` and `extension_sql` macros inside **only that module** disappeared from the linkage in `sql-generator`:

```rust
ana@autonoma:~/git/zombodb/pgx/pgx-tests$ nm -D ../target/debug/libpgx_tests.so | grep __pgx_internals_sql
00000000000f3940 T __pgx_internals_sql_create_complex_shell_type
00000000000f5390 T __pgx_internals_sql_create_complex_type
00000000000aab80 T __pgx_internals_sql_example_file
00000000000aa8d0 T __pgx_internals_sql_extension_sql_dynamic
00000000000aa6e0 T __pgx_internals_sql_extension_sql_table
ana@autonoma:~/git/zombodb/pgx/pgx-tests$ nm -D ../target/debug/sql-generator | grep __pgx_internals_sql
00000000001a04c0 T __pgx_internals_sql_create_complex_shell_type
00000000001a16b0 T __pgx_internals_sql_create_complex_type
```

On develop this works:

```bash
ana@autonoma:~/git/zombodb/pgx/pgx-tests$ nm -D ../target/debug/sql-generator | grep __pgx_internals_sql
00000000001a1940 T __pgx_internals_sql_create_complex_shell_type
00000000001a2ad0 T __pgx_internals_sql_create_complex_type
0000000000166cc0 T __pgx_internals_sql_example_file
0000000000166b30 T __pgx_internals_sql_extension_sql_dynamic
00000000001669a0 T __pgx_internals_sql_extension_sql_table


ana@autonoma:~/git/zombodb/pgx/pgx-tests$ nm -D ../target/debug/libpgx_tests.so | grep __pgx_internals_sql
00000000000f3390 T __pgx_internals_sql_create_complex_shell_type
00000000000f4d80 T __pgx_internals_sql_create_complex_type
0000000000120ff0 T __pgx_internals_sql_example_file
0000000000120e60 T __pgx_internals_sql_extension_sql_dynamic
0000000000120cd0 T __pgx_internals_sql_extension_sql_table
```

**Comment by Hoverbear** [comment]
So this appears to be because of the `format!()` macro usage, as I can use `concat!()` just fine. This is something I'd prefer to resolve before merge...

**Comment by Hoverbear** [comment]
I think we're going to hold this until 0.3.1 until we can wrangle this. We'd like to cut a 0.3.0 soon.

**Comment by Hoverbear** [comment]
I don't really understand why it only happens in pgx-tests...

**Comment by Hoverbear** [comment]
In #441 we significantly refactored some of this code.... There's good news and bad news:

* It will not be a very nice merge -- it may be easier to look at the diff and redo the changes. (edit `pgx_utils/src/sql_entity/graph` only now)
* It should eliminate those bugs I was detecting.

**Comment by Hoverbear** [comment]
@JamesGuthrie is this still of interest to you?

**Comment by workingjubilee** [comment]
Closing this as stale, feel free to reopen this or open a new PR.

**Pull request closed without merging**